enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,194 @@
+---
+title: pre-flight-validations
+authors:
+  - "@mandre"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-18
+last-updated: 2020-05-18
+status: implementable
+---
+
+# Pre-flight validations
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+- Will the validations be run automatically on `cluster create` or will it be
+  an explicit action?
```

Comment:
Preflight checks should be in the default code path. As when it fails it will help communicate why it failed and help the customer fix the issue and run the installer again. 

Also there should be an explicit way to skip preflight checks when required. Sometimes customers might want to skip it because of specific environment requirement. For example a customer trying OpenShift installation in a new public cloud which we do not support and the preflight check fails on the last preflight check. But customer is fine with installing it and then manually fixing the thing which made the preflight check fail.

Reply:
While I agree that in general we'll want the validations to run by default as it is done today, everyone might not be happy running all validations if it makes the deployment significantly longer, and we may want to explicitly enable them instead. I'm thinking in particular about the extra checks to validate the cloud performance or the ability to pull container images.

Agreed on the ability to bypass checks. Depending on how we implement validations, either turn off validations, or make their failures non-fatal.

Reply:
The pre-flight checks should enough value that it should be in the default path. However if you have checks which take longer time and you do not want them to be enabled by default them you can the pre-flight checks in to two parts. Once part which should be run everytime and some extra tests which can be opted to run.

Reply:
But ideally pre-flight checks should not take a long time to run as it should not make any changes to the platform and it adds so much value that it should run every time installer is being run.

Reply:
I guess the prechecks should be optional, but encouraged to be executed. If you have a testing environment where you usually deploy OCP, it is pointless to run those prechecks every time. So I think having a `--skip-prechecks` flag where if not present, the prechecks will be executed. Also, the prechecks shall be able to be executed by themselves, like `openshift-install run-prechecks` or something. My 2 cents.

Reply:
So right now, I'm leaning towards splitting the pre-flight validations into 2 buckets:
- the core validations, enabled all the time - this includes all of the current validations in the installer.
- the extra validations, enabled on demand, for more involved checks requiring to boot a node. The performance validations are a good example of such checks, where it can take quite some time, especially on BM. 

Now, I'm not sure what the best way to enable the extra validations is... It depends if we want to perform the installation or not after running the extra validations:
- a `--dry-run` flag runs all the validation but not performing the installation
- a `--with-extra-validations` or similar flag, enabling the extra validations and performing the installation.

Reply:
I think I've captured this conversation in my [latest patch](https://github.com/openshift/enhancements/pull/346/commits/aef64f011bd4c0db89bf4aea57dfb204d32361b9). Please check.